server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldFetcher.java

```diff
@@ -115,23 +121,31 @@ private static FieldFetcher create(SearchExecutionContext context,
         }
 
         CharacterRunAutomaton unmappedFieldsFetchAutomaton = null;
-        if (unmappedFetchPattern.isEmpty() == false) {
+        Map<Boolean, List<String>> partitions = unmappedFetchPattern.stream()
+            .collect(Collectors.partitioningBy((s -> Regex.isSimpleMatchPattern(s))));
+        List<String> unmappedWildcardPattern = partitions.get(true);
+        List<String> unmappedConcreteFields = partitions.get(false);
```

Comment:
Maybe we could add a short comment about why we are doing this? It is not so obvious, as it addresses a bit of an edge case (specifying many concrete field paths, but also requesting `include_unmapped: true`).